### PR TITLE
Template Metaprogramming Bits.

### PR DESCRIPTION
**Pull request by dkorolev**
To be used by `Sherlock` & master `Yoda`.

### CODE REVIEW

**Comment by dkorolev** [comment]
(Sorry, I can't think of a better name than `pod_constref` now. But it's terrible. Suggestions very welcome!)

**Comment by mzhurovich** [comment]
`cref_not_pod`?

**Comment by dkorolev** [comment]
`byval_or_by_cref` ?

**Comment by dkorolev** [comment]
`copy_less` ?

And I'm in Viber/Skype!

**Comment by dkorolev** [comment]
@mzhurovich : Made the changes we discussed, touched the rest of `Bricks` to use the new `rmref` thing, renamed the unrenamed magic into `copy_free`, and a few minor things.

PTAL!

Thanks,
Dima

File: net/http/impl/server.h [hunk 1]
```diff
@@ -456,7 +457,7 @@ class HTTPServerConnection final {
           connection_.BlockingWrite("0", true);
           connection_.BlockingWrite(kCRLF, false);
           connection_.BlockingWrite(kCRLF, false);  // We should send CRLF twice.
-        } catch (const std::exception& e) {  // LCOV_EXCL_LINE
+        } catch (const std::exception& e) {         // LCOV_EXCL_LINE
```
Comment (mzhurovich) [inline net/http/impl/server.h]: Funny indent :)

File: template/docu/docu_05variadic_02.cc [hunk 2]
```diff
@@ -74,23 +75,95 @@ TEST(Variadic, Filter) {
...
+TEST(TemplateMetaprogramming, Reduce) {
   struct A { static std::string s() { return "A"; } };
   struct B { static std::string s() { return "B"; } };
   struct C { static std::string s() { return "C"; } };
   EXPECT_EQ("(A+(B+C))",
-            (bricks::variadic::reduce<concatenate_s, std::tuple<A, B, C>>::s()));
+            (bricks::metaprogramming::reduce<concatenate_s, std::tuple<A, B, C>>::s()));
 }
     
   // Combine.
```
Comment (mzhurovich) [inline template/docu/docu_05variadic_02.cc]: Indent of the comments here and below looks a bit strange.

File: net/http/impl/server.h (hunk 1 above)
Comment (dkorolev) [inline net/http/impl/server.h]: That's what `clang-format` does. I used to fix it back for a few times, decided to not fight the inevitable this time :-)

File: template/docu/docu_05variadic_02.cc (hunk 2 above)
Comment (dkorolev) [inline template/docu/docu_05variadic_02.cc]: This one is for `make docu`: Without two leading spaces this comment line won't make it to `README.md`.

Whereas for example, the test name doesn't have to be there, and thus is not indented. Agree it looks a bit odd, but does the job well so far.

Comment (mzhurovich) [inline template/docu/docu_05variadic_02.cc]: Ah, right! Forgot about that :)

**Comment by mzhurovich** [comment]
@dkorolev LGTM :+1: 
One more thing - you renamed the directory, but left `ifdef`'s and `docu-`filenames as `variadic`. Is it intentionally?

**Comment by dkorolev** [comment]
Nope, fixing, one minute ...

**Comment by dkorolev** [comment]
@mzhurovich : Thanks, PTAL! :-)

**Comment by mzhurovich** [comment]
@dkorolev `docu_05variadic*` -> `docu_05template*`?

**Comment by dkorolev** [comment]
:+1: Done, I thought `metaprogramming` fits a bit better.

**Comment by mzhurovich** [comment]
@dkorolev +1! Merging.

**Comment by dkorolev** [comment]
Thanks!

**Pull request merged**